[Fixes]: Review comments and fixes

MuhammedZamroodh opened the pull request:
[Description]:

- Fixed use of neop instruction (used nop instead)
- Doc updates in arch and board
- Added optimisation levels
- ISR vector number lost on premption, fixed by adding
- workaround for for alternate vec number calculation

Udhayanandhan-Jayakumar committed:
Fixes: Fix ISR premption

[Description]:
        Fixed use of neop instruction (used nop instead)
        Doc fixes in arch and board
        Added optimisation levels
        ISR vector number lost on premption, fixed by adding
        workaround for for alternate vec number calculation

Signed-off-by: Muhammed Zamroodh <muhammed.zamroodh@ignitarium.com>

pradeepbudagutta commented on boards/microchip/dspic33/dspic33a_curiosity/doc/index.rst:
```diff
@@ -130,6 +130,39 @@ Building
+``cmake/compiler/xcdsc/compiler_flags.cmake`` by mapping Zephyr's generic
+properties to concrete compiler flags. In this file:
+
+* ``optimization_speed`` uses ``-O2``.
```
For speed optimization, it should be -o3

cawilkie replied:
It can go either way.  I see many projects that use -O2.  Shouldn't this be client preference?

MuhammedZamroodh replied:
We do not have license for building with O3,
We all agreed on going ahead with O2 in a mail conversation

pradeepbudagutta commented on boards/microchip/dspic33/dspic33a_curiosity/doc/index.rst:
```diff
@@ -130,6 +130,39 @@ Building
+- ``CONFIG_SIZE_OPTIMIZATIONS_AGGRESSIVE`` switches to ``optimization_size_aggressive``,
+  which adds the ``-Os`` mapping described in ``compiler_flags.cmake``. This
+  level is only available when your XC-DSC compiler is licensed.
+- ``CONFIG_SPEED_OPTIMIZATIONS`` selects ``optimization_speed`` (``-O2``).
```
``optimization_speed`` (``-O3``)

MuhammedZamroodh replied:
We do not have license for building with O3,
We all agreed on going ahead with O2 in a mail conversation

pradeepbudagutta commented on cmake/compiler/xcdsc/compiler_flags.cmake:
```diff
@@ -2,10 +2,10 @@
 # Optimization for debug builds
 set_compiler_property(PROPERTY optimization_debug -O1)
-set_compiler_property(PROPERTY optimization_speed -O1)
+set_compiler_property(PROPERTY optimization_speed -O2)
```
set_compiler_property(PROPERTY optimization_speed -O3)

MuhammedZamroodh replied:
We do not have license for building with O3,
We all agreed on going ahead with O2 in a mail conversation

pradeepbudagutta replied:
This is not related to license and testing but mapping the "speed" option to -o3 which is correct.

pradeepbudagutta replied:
default should be o2 as per mail conversation, but the option of speed must be mapped to -o3

cawilkie commented on drivers/timer/mchp_dspic33_timer.c:
Please write clean inline assembly, this code clobbers w0 without telling the compiler.  Which may be okay iff the call is not inlined and no-body adds any code to it (ie, its not maintainable).

void arch_busy_wait(uint32_t usec_to_wait)
{
	usec_to_wait <<=3;
	__asm__ volatile("repeat.w %0\n\t"
			 "nop\n\n\t" : /* outputs */ : /* inputs */ "=r"(usec_to_wait));
}

MuhammedZamroodh replied:
@cawilkie , sounds good
I have made the changes and pushed it

MuhammedZamroodh committed:
timer: Fix busywait inline assembly

Make inline assembly portable/maintainable

Signed-off-by: Muhammed Zamroodh <muhammed.zamroodh@ignitarium.com>

harishagari commented on soc/microchip/dspic33/dspic33a/soc.c:
```diff
@@ -6,6 +6,8 @@
+/* SOC level init for dspic33a family of chips
+ */
 void soc_early_init_hook(void)
 {
```
Intention of my previous comment(on previous PR) was to add the comment inside function to satisfy the MISRA violations about empty function body.

MuhammedZamroodh committed:
cmake: Fix compilation flag for speed

Change the optimisation flag for speed from o2 to o3
(o3 needs licensed version of toolchain)

Signed-off-by: Muhammed Zamroodh <muhammed.zamroodh@ignitarium.com>
